partiql-eval/src/test/kotlin/org/partiql/eval/internal/WindowTests.kt

```diff
@@ -0,0 +1,368 @@
+package org.partiql.eval.internal
+
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.parallel.Execution
+import org.junit.jupiter.api.parallel.ExecutionMode
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.MethodSource
+import org.partiql.eval.Mode
+import org.partiql.spi.value.Datum
+import org.partiql.spi.value.Field
+
+/**
+ * This test file tests Common Table Expressions.
+ */
+class WindowTests {
+
+    @ParameterizedTest
+    @MethodSource("successTestCases")
+    @Execution(ExecutionMode.CONCURRENT)
+    fun successTests(tc: SuccessTestCase) = tc.run()
+
+    @ParameterizedTest
+    @MethodSource("failureTestCases")
+    @Execution(ExecutionMode.CONCURRENT)
+    fun failureTests(tc: FailureTestCase) = tc.run()
+
+    /**
+     * This is used just for debugging purposes.
+     */
+    @Test
+    fun developmentTest() {
+        val failingIndex = 3
+        successTestCases()[failingIndex - 1].run()
+    }
+
+    companion object {
+
+        private class Employee(
+            val id: Int,
+            val name: String,
+            val department: String,
+            val age: Int,
+            val partner: String?
+        ) {
+            fun toDatum(): Datum {
+                val fields = listOfNotNull(
+                    Field.of("id", Datum.integer(id)),
+                    Field.of("name", Datum.string(name)),
+                    Field.of("department", Datum.string(department)),
+                    Field.of("age", Datum.integer(age)),
+                    partner?.let { Field.of("partner", Datum.string(it)) }
+                )
+                return Datum.struct(fields)
+            }
+        }
+
+        private val employees = listOf(
+            Employee(0, "Jacob", "Marketing", 40, "Alexa"),
+            Employee(1, "Marcus", "Sales", 28, "Gary"),
+            Employee(2, "Shelly", "Research", 35, "Michael"),
+            Employee(3, "Alexa", "Research", 30, null),
+            Employee(4, "Raghavan", "Sales", 28, "Yi"),
+            Employee(5, "Yi", "Research", 25, "Raghavan"),
+            Employee(6, "Megan", "Marketing", 32, null),
+            Employee(7, "Amanda", "Research", 32, null),
+            Employee(8, "Samantha", "Sales", 29, "Mason"),
+            Employee(9, "Mason", "Research", 30, "Samantha")
+        )
+
+        private val globals = listOf(
+            Global(
+                name = "employee",
+                value = Datum.bag(employees.map { it.toDatum() })
+            )
+        )
+
+        private const val FALLBACK: String = "UNKNOWN"
+
+        @JvmStatic
+        fun successTestCases() = listOf(
```

Comment:
A couple more test I think would be beneficial here: 

Using window function with group by: [ref](https://github.com/partiql/partiql-lang-kotlin/blob/c26536c149bb58d081b4f41544f1ebfd532f26a0/lang/src/test/kotlin/org/partiql/lang/eval/builtins/windowFunctions/WindowFunctionTests.kt#L111-L127)

Using window function with subquery: [ref](https://github.com/partiql/partiql-lang-kotlin/blob/c26536c149bb58d081b4f41544f1ebfd532f26a0/lang/src/test/kotlin/org/partiql/lang/eval/builtins/windowFunctions/WindowFunctionTests.kt#L149-L157)

Window order does not affect outer order: [ref](https://github.com/partiql/partiql-lang-kotlin/blob/c26536c149bb58d081b4f41544f1ebfd532f26a0/lang/src/test/kotlin/org/partiql/lang/eval/builtins/windowFunctions/WindowFunctionTests.kt#L91-L108)